src/man/man1/mbm_makedatalist.1
.TH mbm_makedatalist 1 "22 June 2022" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbm_makedatalist\fP \- Macro to generate a datalist of the swath
files in a specified directory.

.SH VERSION
Version 5.0

.SH SYNOPSIS
\fBmbm_makedatalist\fP [\fB\-B\fP\fIsize\fP \fB\-F\fP\fIformat\fP \fB\-I\fP\fIdirectory\fP \fB-L\fP
\fB\-O\fP\fIdatalist\fP \fB-P\fP \fB-S\fP\fIsuffix\fP \fB\-T -H \-V\fP]

.SH DESCRIPTION
\fBMbm_makedatalist\fP is a macro to generate an MB-System datalist file
referencing all identifiable swath files in the specified target directory.

Datalists are fundamental structures in \fBMB-System\fP workflows because they
allow programs to operate on sets of swath data files.
Datalist files are text lists of swath data files and their format ids with each
file entry taking up a single line. These lists may contain references to other
datalists, making them recursive. Datalists may also contain comments and parsing
directives that, for example, determine whether parsing returns references to
raw or processed data files. See the \fBMB-System\fP manual page for details
on the format and structure of datalists.

This macro is used to easily construct a datalist referencing swath files in a
directory, often as part of setting up the \fBMB-System\fP processing environment
for a new seafloor mapping dataset. By default, \fBmbm_makedatalist\fP tests
all of the filenames in the local directory for adherence to the \fBMB-System\fP
swath file naming convention (see the \fBmbsystem\fP manual page for details),
and makes an output datalist including all of the files that can be recognized
as swath data. This can include files with suffixes such as ".mb88", that are
recognized as format 88 (Reson s7k) files, and files with recognized vendor
format suffixes, such as "_raw.all" or ".all" for old and recent Kongsberg
multibeam data, respectively. The suffix of the files to be considered can be
specified with the \fB-S\fP\fIsuffix\fP option; under most circumstances this is
the best approach as it prevents \fBmbm_makedatalist\fP from including files that
are not intended for inclusion by the user yet satisfy one of the many recognized
naming conventions. By default, \fBmbm_makedatalist\fP infers the format id
of included files using the naming conventions. The format id can be specified
using the \fB-F\fP\fIformat\fP option.

\fBMB-System\fP differentiates between raw and processed files by inserting the
letter "p" into the processed file name immediately before the filename suffix.
For instance, for Reson 7k data the raw files will have file suffix of *.mb88,
and the processed files generated by \fBmbprocess\fP will end with *p.mb88.
When using \fBmbm_makedatalist\fP to generate a datalist by specifying a suffix,
processed files can be excluded from the datalist by also specifying the \fB-P\fP
option. Thus it is possible to generate a datalist of raw files in a directory
containing both raw and processed file.

The output datalist will be placed in the current working
directory. The output datalist is named datalist.mb-1
by default, but can be specified with the \fB\-O\fP option.
If no target directory is specified with the \fB\-I\fP
option, then the current working directory is used. In order
for swath files to be recognized and included in the datalist,
they must conform to the MB-System swath file naming convention
(see the \fBmbsystem\fP manual page for details).

Most seafloor mapping sonars and datalogging packages produce original swath
data files with names constructed so that standard directory listings are in
time order. However, Kongsberg multibeam data filenames usually start with a survey
file number that restarts at 0001 whenever datalogging is stopped and restarted.
Consequently, when data from multiple instances of logging are present in a
single directory, the simple listing of files will not in general be in time
order. If the file suffix is declared to be ".all", ".ALL", ".kmall", or ".KMALL" using the \fB-S\fP
option, then \fBmbm_makedatalist\fP will attempt to sort the swath files into
time order before generating the final datalist. This behavior can be disabled
by specifying the \fB-T\fP option. Kongsberg dataloggers often create junk data
files with names such as "9999.all" or "9999.ALL"; these filenames are ignored.

.SH MB-SYSTEM AUTHORSHIP
David W. Caress
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes
.br
  Center for Coastal and Ocean Mapping
.br
  University of New Hampshire
.br
Christian do Santos Ferreira
.br
  MARUM - Center for Marine Environmental Sciences
.br
  University of Bremen

.SH OPTIONS
.TP
.B \-B
\fIsize\fP
.br
This option sets a size threshold in kilobytes below which a file will be
ignored and not included in the output datalist file. 
.TP
.B \-F
\fIformat\fP
.br
By default, \fBmbm_makedatalist\fP infers the format id
of included files using the naming conventions. This option specifies the
format id that is assigned.
.TP
.B \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-I
\fIdirectory\fP
.br
Sets the path of the target directory.
.TP
.B \-L
.br
Causes the last file in a directory listing to be omitted from the output datalist.
The intent of this option is to exclude the most recent file in an actively data
logging context, since the most recent file is still being logged. This presumes
that filenames sort in time order as listed by the \fBls\fP command, an assumption
that does not hold for all file naming conditions or if the \fB-T\fP option has
been used to disable time ordering.
.TP
.B \-O
\fIdatalist\fP
.br
Sets the name of the output datalist file. Default: datalist.mb-1
.TP
.B \-P
.br
Suppresses the inclusion of processed files in the output datalist (e.g.
*p.mb88).
.TP
.B \-S
\fIsuffix\fP
.br
Sets the filename suffix of swath files considered for inclusion in the output
datalist. Under most circumstances using this option explicitly is the best
approach as it prevents \fBmbm_makedatalist\fP from including files that
are not intended for inclusion by the user yet satisfy one of the many recognized
naming conventions.
If the file suffix declared with this option corresponds to one of the formats
associated with Kongsberg multibeam sonar data, then \fBmbm_makedatalist\fP will
attempt to sort the swath files into time order before generating the final datalist.
This behavior applies to files with the ".all" or ".ALL" suffix used for raw logged
Kongsberg files, and also to files with suffixes following the \fBMB-System\fP
convention (e.g. ".mb56", ".mb57", ".mb58", ".mb59" suffixes). This behavior is
disabled by the \fB-T\fP option.
.TP
.B \-T
.br
By default, when the \fB-S\fP option is used to specify a file suffix corresponding
to Kongsberg multibeam data, \fBmbm_makedatalist\fP attempts to sort the data files
into time order before outputting the datalist. This option disables that behavior.
.TP
.B \-V
The \fB\-V\fP option causes \fBmbm_makedatalist\fP to print out status messages.

.SH EXAMPLES
Suppose we have a set of swath files in a directory and we
want to construct a datalist referring to those files. A simple
directory listing shows four files:
.br
 	20051019_154030.mb88
 	20051019_162540.mb88
 	20051019_171756.mb88
 	junk.txt
.br
of which three are identifiable as swath files because of the
".mb88" file suffix. Running \fBmbm_makedatalist\fP in this
directory with no arguments other than verbosity:
.br
 	mbm_makedatalist \-V
.br
produces a datalist file named datalist.mb-1. The contents of
this file are:
.br
 	20051019_154030.mb88 88
 	20051019_162540.mb88 88
 	20051019_171756.mb88 88
.br
where the file junk.txt has been ignored.

.SH SEE ALSO
\fBmbsystem\fP(1), \fBmbdatalist\fP(1), \fBmbinfo\fP(1)

.SH BUGS
Only fake bugs here.


NEWS
-------------------------------------------------------------------------------
MB-SYSTEM NEWS FILE:
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

In the GNU package standard, the NEWS file lists the history of user-visible
changes to a package. For the the MB-System open source software package for the
processing and display of swath sonar data, this information is included in the
ChangeLog file. Both the NEWS file and the ChangeLog file are located at the top
of the MB-System source code distribution directory structure.

-------------------------------------------------------------------------------


trunk/src/macros/mbm_makedatalist.pl
eval '(exit $?0)' && eval 'exec perl -S $0 ${1+"$@"}'
                    & eval 'exec perl -S $0 $argv:q'
                         if 0;
#--------------------------------------------------------------------
#    The MB-system:	mbm_makedatalist.perl	27/1/2006
#    $Id$
#
#    Copyright (c) 2006-2012 by
#    D. W. Caress (caress@mbari.org)
#      Monterey Bay Aquarium Research Institute
#      Moss Landing, CA
#    and D. N. Chayes (dale@lamont.ldgo.columbia.edu)
#      Lamont-Doherty Earth Observatory
#      Palisades, NY  10964
#
#    See README file for copying and redistribution conditions.
#--------------------------------------------------------------------
#
# Command:
#   mbm_makedatalist
#
# Purpose:
#   Macro to generate an MB-System datalist file referencing all
#   identifiable swath files in the specified directory. If no directory
#   is specified with the -I option, then the current directory is used.
#   The resulting datalist will be named datalist.mb-1 by default.
#
# Basic Usage:
#   mbm_makedatalist [-Idirectory -Odatalist -H -V]
#
# Author:
#   David W. Caress
#   Monterey Bay Aquarium Research Institute
#   Moss Landing, CA 95039
#   January 27, 2006
#
# Version:
#   $Id$
#
# Revisions:
#   $Log: mbm_makedatalist.perl,v $
#   Revision 5.1  2006/04/11 19:12:06  caress
#   Macro now handles data in remote directories.
#
#   Revision 5.0  2006/02/03 21:06:18  caress
#   Initial version of macro to generate a datalist of all recognizable swath files in a directory.
#
#
#
$program_name = "mbm_makedatalist";

# Deal with command line arguments
$command_line = "@ARGV";
&MBGetopts('HhI:i:O:o:Vv');

$help =    		($opt_H || $opt_h);
$directory = 		($opt_I || $opt_i || ".");
$datalist =    		($opt_O || $opt_o || "datalist.mb-1");
$verbose = 		($opt_V || $opt_v);

# print out help message if required
if ($help)
	{
	print "\n$program_name:\n";
	print "\nVersion: $Id$\n";
        print "Macro to generate an MB-System datalist file referencing all\n";
        print "identifiable swath files in the specified directory. If no directory\n";
        print "is specified with the -I option, then the current directory is used.\n";
        print "The resulting datalist will be named datalist.mb-1 by default.\n";

        print "\nBasic Usage:\n\tmbm_makedatalist [-Idirectory -Odatalist -H -V]\n";
	exit 0;
	}

# tell the world we got started
if ($verbose)
	{
	print "\nRunning $program_name...\n";
	}

# get list of files in specified directory
@filelist = `ls -1 $directory`;

# loop over the list of files counting identifiable swath files
$count = 0;
foreach $file (@filelist)
	{
	chop $file;
	$format = `mbformat -L -I $file`;
	chop $format;
	if ($format != 0 && ($file ne $datalist || $directory ne "."))
		{
		$count++;
		if ($directory ne ".")
			{
			$path = "$directory/$file";
			}
		else
			{
			$path = $file;
			}
		push(@outfilelist, $path);
		push(@outformatlist, $format);
		}
	}

# open output datalist
if ($count > 0)
	{
	if (!open(DATALIST,">$datalist"))
		{
		die "\n$program_name:\nUnable to output output datalist file $datalist\nExiting...\n";
		}

	# loop over the list of files
	for ($i = 0; $i < scalar(@outfilelist); $i++)
		{
		print DATALIST "$outfilelist[$i] $outformatlist[$i]\n";
		$count++;
		}

	# close datalist
	close DATALIST;
	}

# done
exit 0;

#-----------------------------------------------------------------------
# This version of Getopts has been augmented to support multiple
# calls to the same option. If an arg in argumentative is followed
# by "+" rather than ":",  then the corresponding scalar will
# be concatenated rather than overwritten by multiple calls to
# the same arg.
#
# Usage:
#      do Getopts('a:b+c'); # -a takes arg, -b concatenates args,
#			    # -c does not take arg. Sets opt_* as a
#                           # side effect.

sub MBGetopts {
    local($argumentative) = @_;
    local(@args,$_,$first,$rest);
    local($errs) = 0;

    @args = split( / */, $argumentative );
    while(@ARGV && ($_ = $ARGV[0]) =~ /^-(.)(.*)/) {
	($first,$rest) = ($1,$2);
	$pos = index($argumentative,$first);
	if($pos >= $[) {
	    if($args[$pos+1] eq ':') {
		shift(@ARGV);
		if($rest eq '') {
		    ++$errs unless @ARGV;
		    $rest = shift(@ARGV);
		}
		eval "\$opt_$first = \$rest;";
		eval "\$flg_$first = 1;";
	    }
	    elsif($args[$pos+1] eq '+') {
		shift(@ARGV);
		if($rest eq '') {
		    ++$errs unless @ARGV;
		    $rest = shift(@ARGV);
		}
		if (eval "\$opt_$first") {
		    eval "\$opt_$first = \$opt_$first
				. \":\" . \$rest;";
		}
		else {
		    eval "\$opt_$first = \$rest;";
		}
		eval "\$flg_$first = 1;";
	    }
	    elsif($args[$pos+1] eq '%') {
		shift(@ARGV);
		if($rest ne '') {
		    eval "\$opt_$first = \$rest;";
		}
		else {
		    $rest = $ARGV[0];
		    ($one) = $rest =~ /^-(.).*/;
		    $pos = index($argumentative,$one);
		    if(!$one || $pos < $[) {
			eval "\$opt_$first = \$rest;";
			shift(@ARGV);
		    }
		}
		eval "\$flg_$first = 1;";
	    }
	    else {
		eval "\$opt_$first = 1";
		eval "\$flg_$first = 1;";
		if($rest eq '') {
		    shift(@ARGV);
		}
		else {
		    $ARGV[0] = "-$rest";
		}
	    }
	}
	else {
	    print STDERR "Unknown option: $first\n";
	    ++$errs;
	    if($rest ne '') {
		$ARGV[0] = "-$rest";
	    }
	    else {
		shift(@ARGV);
	    }
	}
    }
    $errs == 0;
}
#-----------------------------------------------------------------------
